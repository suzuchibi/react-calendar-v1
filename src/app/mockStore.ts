import configureMockStore from 'redux-mock-store';
const mock = configureMockStore();

export const mockStore = mock({
  drawer: false,
  today: 1626015600000,
  modal: {
    isActive: false,
    mode: '0',
  },
  form: {
    id: '0',
    isDuplicate: false,
  },
  cal: {
    current: 1625065200000,
    holidays: [
      {
        date: '2021-07-22',
        summary: '海の日',
      },
    ],
    status: 'loaded',
  },
  calPoper: {
    anchorEl: null,
  },
  schedule: {
    holidays: [],
    selected: 1626015600000,
    status: 'loaded',
  },
  datas: [
    {
      id: '1',
      title: 'タイトル1',
      coment: 'コメント1',
      start: 1626139200000,
      end: 1626141600000,
    },
  ],
  screen: false,
  snack: {
    open: false,
    msg: '',
  },
});
