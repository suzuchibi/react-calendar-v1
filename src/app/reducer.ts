import { combineReducers } from 'redux';
import isDrawerReducer from '../features/side/isDrawerReducer';
import todayReducer from '../features/today/todayReducer';
import modalReducer from '../features/modal/reducer';
import formReducer from '../features/form/reducer';
import calendarReducer from '../features/calendar/calendarReducer';
import calendarPoperReducer from '../features/dialog/calendarPoper/reducer';
import scheduleReducer from '../features/scheduler/reducers/scheduleReducer';
import scheduleDatasReducer from '../features/scheduler/reducers/datasReducer';
import fullScreenDialogReducer from '../features/dialog/fullScreen/reducer';
import snackReducer from '../features/snack/snackReducer';

const rootReducer = combineReducers({
  drawer: isDrawerReducer,
  today: todayReducer,
  modal: modalReducer,
  form: formReducer,
  cal: calendarReducer,
  calPoper: calendarPoperReducer,
  schedule: scheduleReducer,
  datas: scheduleDatasReducer,
  screen: fullScreenDialogReducer,
  snack: snackReducer,
});

export type RootState = ReturnType<typeof rootReducer>;
export default rootReducer;
