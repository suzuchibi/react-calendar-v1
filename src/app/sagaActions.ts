import { Holidays } from '../features/calendar/types';

const GOOGLE_CALENDAR_API = 'app/googleCalendarAPI';
const GOOGLE_CALENDAR_API_SUCESS = 'app/googleCalendarApiSuccess';
const GOOGLE_CALENDAR_API_ERROR = 'app/googleCalendarApiError';

export interface Props {
  mode: string;
  stamp: number;
}

export interface GoogleCalendarApiProps {
  type: typeof GOOGLE_CALENDAR_API;
  payload: Props;
}

interface HolidayProps {
  mode: string;
  result: Array<Holidays>;
}

export interface GoogleCalendarApiSuccessProps {
  type: typeof GOOGLE_CALENDAR_API_SUCESS;
  payload: HolidayProps;
}

export interface GoogleCalendarApiErrorProps {
  type: typeof GOOGLE_CALENDAR_API_ERROR;
  payload: {
    mode: string;
  };
}

export type ActionTypes =
  | GoogleCalendarApiProps
  | GoogleCalendarApiSuccessProps
  | GoogleCalendarApiErrorProps;

/**
 * google calendar API saga start
 * @param {object} payload
 */
export const gApiSaga = (payload: Props): ActionTypes => ({
  type: GOOGLE_CALENDAR_API,
  payload,
});

/**
 * Success
 * @param {object} payload
 */
export const gApiSuccess = (payload: HolidayProps): ActionTypes => ({
  type: GOOGLE_CALENDAR_API_SUCESS,
  payload,
});

/**
 * Error
 * @param {object} payload
 */
export const gApiError = (payload: { mode: string }): ActionTypes => ({
  type: GOOGLE_CALENDAR_API_ERROR,
  payload,
});
