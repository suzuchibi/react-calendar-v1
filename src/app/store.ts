import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';

import rootSaga from './sagas';
import rootReducer from './reducer';

const saga = createSagaMiddleware();
const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(saga))
);
saga.run(rootSaga);

export default store;
