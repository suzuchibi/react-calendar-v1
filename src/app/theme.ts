import { createMuiTheme } from '@material-ui/core/styles';

export const theme = createMuiTheme({
  palette: {
    background: {
      default: '#fff',
    },
  },
});
