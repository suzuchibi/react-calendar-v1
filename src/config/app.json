{
  "drawerWidth": 260,
  "timeLineWidth": 60,
  "startOrder": 10,
  "lastOrder": 19,
  "dayOfTheWeek": ["日", "月", "火", "水", "木", "金", "土"],
  "gapiCal": {
    "key": "AIzaSyCLqi5tQwWw8qEGAWc9NC1cEU78C4EAWkM",
    "address": "ja.japanese#holiday@group.v.calendar.google.com",
    "gurl": "https://www.googleapis.com/calendar/v3/calendars/"
  }
}