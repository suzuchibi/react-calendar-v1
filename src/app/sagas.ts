import { put, call, takeEvery } from 'redux-saga/effects';
import { CallEffect, PutEffect } from 'redux-saga/effects';
import { Props as SagaProps } from './sagaActions';
import { Holidays } from '../features/calendar/types';
import GAPI from '../api/googleCalendarApi';

import config from '../config/app.json';
import {
  getYear,
  getMonth,
  getDate,
  startOfMonth,
  endOfMonth,
  add,
} from 'date-fns';

export interface ApiProps {
  type: string;
  payload: SagaProps;
}
export interface SuccessProps {
  type: string;
  payload: {
    mode: string;
    result: Array<Holidays>;
  };
}
export interface ErrorProps {
  type: string;
  payload: {
    mode: string;
  };
}
export type SagaTypes = CallEffect | PutEffect;

/**
 * SetDateText タイムゾーンのテキスト化
 * @param {Date | number} e
 * @return {String}
 */
const setTimeText = (e: Date | number) => {
  const pad = (t: number) => ('00' + t).slice(-2);
  return `${getYear(e)}-${pad(getMonth(e) + 1)}-${pad(getDate(e))}T00:00:00Z`;
};

/**
 * GetTimeZone 開始と終了の日付取得
 * @param {number} stamp
 * @param {String} mode
 * @return {Object<{start: String, end: String}>}
 */
const getTimeZone = (stamp: number, mode: String) => {
  switch (mode) {
    case 'cal':
      const s = startOfMonth(stamp);
      const e = endOfMonth(stamp);
      return {
        start: setTimeText(s),
        end: setTimeText(e),
      };
    case 'schedule':
      const dateAddSix = add(stamp, { days: 7 });
      return {
        start: setTimeText(stamp),
        end: setTimeText(dateAddSix),
      };
    case 'all':
      const sAll = startOfMonth(stamp);
      const end = endOfMonth(stamp);
      const endAddSix = add(end, { days: 6 });
      return {
        start: setTimeText(sAll),
        end: setTimeText(endAddSix),
      };
    default:
      return {
        start: '',
        end: '',
      };
  }
};

// ここからジェネレーター

/**
 * Success 成功した後のディスパッチ処理
 * @param {object} payload
 * @returns {void}
 */
export function* success({
  payload,
}: SuccessProps): Generator<PutEffect, void, SuccessProps> {
  const { mode, result } = payload;
  switch (mode) {
    case 'cal':
      yield put({ type: 'calendar/changeHolidays', payload: result });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      break;
    case 'schedule':
      yield put({ type: 'schedule/changeHolidays', payload: result });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
    case 'all':
      yield put({ type: 'calendar/changeHolidays', payload: result });
      yield put({ type: 'schedule/changeHolidays', payload: result });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
  }
}

/**
 * Error 失敗した後のディスパッチ処理
 * @param {object} payload modeの値が返る
 * @returns {void}
 */
export function* error({
  payload,
}: ErrorProps): Generator<PutEffect, void, string> {
  const { mode } = payload;
  switch (mode) {
    case 'cal':
      yield put({ type: 'calendar/changeHolidays', payload: [] });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      break;
    case 'schedule':
      yield put({ type: 'schedule/changeHolidays', payload: [] });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
    case 'all':
      yield put({ type: 'calendar/changeHolidays', payload: [] });
      yield put({ type: 'schedule/changeHolidays', payload: [] });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
  }
}

/**
 * Generator GoogleCalendarApi
 * @param {object} payload
 * @returns {void}
 */
export function* googleCalendarApiSaga({
  payload,
}: ApiProps): Generator<SagaTypes, void, any> {
  const { mode, stamp } = payload;
  const { start, end } = getTimeZone(stamp, mode);
  const param = `timeMin=${start}&timeMax=${end}&singleEvents=true`;
  const encode = encodeURIComponent(config.gapiCal.address);
  const path = `${config.gapiCal.gurl}${encode}/events?${param}`;
  try {
    // GoogleCalendarAPIコール
    const result = yield call(GAPI.start, path);
    yield put({
      type: 'app/googleCalendarApiSuccess',
      payload: { mode: mode, result: result },
    });
  } catch (err) {
    yield put({ type: 'snac/changeMsg', payload: err });
    yield put({ type: 'snac/changeOpen', payload: true });
    yield put({
      type: 'app/googleCalendarApiError',
      payload: { mode: mode },
    });
  }
}

export default function* rootSaga() {
  yield takeEvery('app/googleCalendarAPI', googleCalendarApiSaga);
  yield takeEvery('app/googleCalendarApiSuccess', success);
  yield takeEvery('app/googleCalendarApiError', error);
}
