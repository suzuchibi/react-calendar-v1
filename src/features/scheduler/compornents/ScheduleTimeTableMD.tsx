import React from 'react';
import config from '../../../config/app.json';
import classNames from 'classnames';
import { makeStyles } from '@material-ui/core/styles';
import blue from '@material-ui/core/colors/blue';
import red from '@material-ui/core/colors/red';
import ScheduleTimeDataMD from './ScheduleTimeDataMD';
import { getYear, getMonth, getDate, getDay } from 'date-fns';

// Types
import { Holidays, Datas } from '../types';
type Props = {
  stamp: number;
  holidays: Array<Holidays>;
  data?: Datas;
  index: Number;
};

/**
 * CSS Styling
 */
const useStyles = makeStyles(() => ({
  timeTableWrap: {
    width: 'calc(100% / 7)',
    margin: 0,
    padding: 0,
  },
  timeTableTitle: {
    height: '48px',
    textAlign: 'center',
    lineHeight: '1.5em',
    margin: 0,
    padding: '0 0 5px 0',
    display: 'flex',
    flexDirection: 'column-reverse',
  },
  timeTableTitleDate: {
    fontSize: '20px',
  },
  timeTableTitleDateDay: {
    fontSize: '12px',
  },
  timeTableHoliday: {
    height: '32px',
    textAlign: 'center',
    borderRight: 'solid 1px #ccc',
    borderBottom: 'solid 1px #ccc',
    display: 'block',
    margin: 0,
    padding: '4px 0 0 0',
  },
  saturday: {
    color: blue[700],
  },
  sunday: {
    color: red[500],
  },
  jHoliday: {
    color: red[500],
  },
  jHolidayName: {
    color: red[500],
    fontSize: '12px',
    fontWeight: 800,
    padding: 0,
  },
  timeTableStage: {
    margin: 0,
    padding: 0,
  },
  timeTableItems: {
    position: 'relative',
    listStyleType: 'none',
    margin: 0,
    padding: 0,
  },
  timeTableItem: {
    height: '120px',
    borderRight: 'solid 1px #ccc',
    borderBottom: 'solid 1px #ccc',
    margin: 0,
    padding: 0,
  },
}));

/**
 * initTimeLine
 * @return {array}
 */
const initTimeLine = () => {
  const array = [];
  const start = config.startOrder;
  const end = config.lastOrder;
  for (let i = 0; i <= end - start; i++) {
    array.push(start + i);
  }
  return array;
};

/**
 * Check Japanese Holiday
 * @param {number} stamp
 * @param {Array} holidays
 */
const isJHoliday = (stamp: number, holidays: Array<Holidays>) => {
  let name = null;
  const year = getYear(stamp);
  const month = getMonth(stamp) + 1;
  const day = getDate(stamp);
  const zeroPad = (e: number) => ('00' + e).slice(-2);
  const target = `${year}-${zeroPad(month)}-${zeroPad(day)}`;
  holidays.forEach((v) => {
    if (v.date) {
      if (v.date === target) name = v.summary;
    }
  });
  return name;
};

/**
 * Render
 */
function ScheduleTimeTableMD(props: Props) {
  const classes = useStyles();
  const timeLines = initTimeLine();
  const dayOfTheWeek = config.dayOfTheWeek;
  const jHoliday = isJHoliday(props.stamp, props.holidays);
  const month = getMonth(props.stamp) + 1;
  const date = getDate(props.stamp);
  const day = getDay(props.stamp);
  const color = (sum: Number, i: Number) => {
    const count = Math.floor(Number(i) / 3);
    return 3 < sum ? Number(i) - 3 * count : sum;
  };

  return (
    <dl className={classes.timeTableWrap}>
      <dt
        className={classNames([classes.timeTableTitle], {
          [classes.sunday]: day === 0,
          [classes.saturday]: day === 6,
          [classes.jHoliday]: jHoliday !== null,
        })}
      >
        <span className={classes.timeTableTitleDate}>
          {month}/{date}
        </span>
        <span className={classes.timeTableTitleDateDay}>
          {dayOfTheWeek[day]}
        </span>
      </dt>
      <dd className={classes.timeTableHoliday}>
        <span className={classes.jHolidayName}>{jHoliday}</span>
      </dd>
      <dd className={classes.timeTableStage}>
        <ul className={classes.timeTableItems}>
          {timeLines.map((v) => {
            return <li className={classes.timeTableItem} key={v} />;
          })}
          {props.data &&
            props.data.map((v, i) => {
              return (
                <ScheduleTimeDataMD
                  data={v}
                  color={color(Number(props.index) + i, i)}
                  key={i}
                />
              );
            })}
        </ul>
      </dd>
    </dl>
  );
}

export default ScheduleTimeTableMD;
