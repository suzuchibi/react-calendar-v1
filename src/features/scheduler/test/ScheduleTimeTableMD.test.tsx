import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import ScheduleTimeTableMD from '../compornents/ScheduleTimeTableMD';

import { mockStore as store } from '../../../app/mockStore';

// Termial
// yarn test src/features/scheduler/test/ScheduleTimeTableMD.test.tsx

describe('<ScheduleTimeWeekMD>', () => {
  const props = {
    stamp: 1626015600000,
    data: [
      {
        id: '3',
        title: 'タイトル3',
        coment: 'コメント３',
        start: 1626235200000,
        end: 1626242400000,
      },
    ],
    holidays: [],
    index: 0,
    key: 0,
  };
  it('スナップショット テスト', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ScheduleTimeTableMD {...props} />
      </Provider>
    );
    // snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
