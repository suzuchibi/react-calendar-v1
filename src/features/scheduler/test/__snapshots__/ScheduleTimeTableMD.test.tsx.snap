// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScheduleTimeWeekMD> スナップショット テスト 1`] = `
<DocumentFragment>
  <dl
    class="makeStyles-timeTableWrap-1"
  >
    <dt
      class="makeStyles-timeTableTitle-2"
    >
      <span
        class="makeStyles-timeTableTitleDate-3"
      >
        7/12
      </span>
      <span
        class="makeStyles-timeTableTitleDateDay-4"
      >
        月
      </span>
    </dt>
    <dd
      class="makeStyles-timeTableHoliday-5"
    >
      <span
        class="makeStyles-jHolidayName-9"
      />
    </dd>
    <dd
      class="makeStyles-timeTableStage-10"
    >
      <ul
        class="makeStyles-timeTableItems-11"
      >
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeTableItem-12"
        />
        <li
          class="makeStyles-timeData-13 makeStyles-color0-15"
          style="height: 239px; top: 360px;"
        >
          <div>
            13:00
          </div>
          <div
            class="makeStyles-timeDataTitle-14"
          >
            タイトル3
          </div>
          <div>
            15:00
          </div>
        </li>
      </ul>
    </dd>
  </dl>
</DocumentFragment>
`;
